language/move-lang/src/parser/syntax.rs

```diff
@@ -392,6 +392,77 @@ fn parse_visibility(tokens: &mut Lexer) -> Result<Visibility, Error> {
         }
     })
 }
+// Parse an attribute value. Either a value literal or a module access
+//      AttributeValue =
+//          <Value>
+//          | <ModuleAccess>
+fn parse_attribute_value(tokens: &mut Lexer) -> Result<AttributeValue, Error> {
+    if tokens.peek() == Tok::NumValue {
+        let loc = current_token_loc(tokens);
+        return Ok(sp(loc, AttributeValue_::NumValue(parse_num(tokens)?)));
+    }
+
+    if let Some(v) = maybe_parse_value(tokens)? {
+        return Ok(sp(v.loc, AttributeValue_::Value(v)));
+    }
+
+    let ma = parse_module_access(tokens, || "attribute name value".to_string())?;
+    Ok(sp(ma.loc, AttributeValue_::ModuleAccess(ma)))
+}
+
+// Parse a single attribute
+//      Attribute =
+//          <Identifier>
+//          | <Identifier> "=" <AttributeValue>
+//          | <Identifier> "(" Comma<Attribute> ")"
+fn parse_attribute(tokens: &mut Lexer) -> Result<Attribute, Error> {
+    let start_loc = tokens.start_loc();
+    let n = parse_identifier(tokens)?;
+    let attr_ = match tokens.peek() {
+        Tok::Equal => {
+            tokens.advance()?;
+            Attribute_::Assigned(n, parse_attribute_value(tokens)?)
+        }
+        Tok::LParen => {
+            let start_loc = tokens.start_loc();
+            tokens.advance()?;
+
+            let args_ = parse_comma_list_after_start(
+                tokens,
+                start_loc,
+                Tok::LParen,
+                Tok::RParen,
+                parse_attribute,
+                "attribute",
+            )?;
+            let end_loc = tokens.previous_end_loc();
+            Attribute_::Parameterized(n, spanned(tokens.file_name(), start_loc, end_loc, args_))
+        }
+        _ => Attribute_::Name(n),
+    };
+    let end_loc = tokens.previous_end_loc();
+    Ok(spanned(tokens.file_name(), start_loc, end_loc, attr_))
+}
+
+// Parse attributes. Used to annotate a variety of AST nodes
+//      Attributes = ("#[" Comma<Attribute> "]")*
```

Comment:
Nit: the NumSign and LBracket are separate tokens, so I suggest writing that as "#" "[" instead of "#["